File: erp/backend/models/vendas.py
from sqlalchemy import String, Integer, ForeignKey, DateTime, Date, Enum, Text, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship
from datetime import datetime, date, timezone
from typing import Optional
import enum
from core.database import Base


class TipoPessoa(str, enum.Enum):
    FISICA = "fisica"
    JURIDICA = "juridica"


class StatusCliente(str, enum.Enum):
    ATIVO = "ativo"
    INATIVO = "inativo"
    BLOQUEADO = "bloqueado"


class StatusPedido(str, enum.Enum):
    RASCUNHO = "rascunho"
    CONFIRMADO = "confirmado"
    FATURADO = "faturado"
    CANCELADO = "cancelado"


class TipoMovEstoque(str, enum.Enum):
    ENTRADA = "entrada"
    SAIDA = "saida"
    RESERVA = "reserva"
    CANCELAMENTO_RESERVA = "cancelamento_reserva"


class Cliente(Base):
    __tablename__ = "clientes"

    id: Mapped[int] = mapped_column(primary_key=True)
    tipo: Mapped[TipoPessoa] = mapped_column(Enum(TipoPessoa))
    nome: Mapped[str] = mapped_column(String(255), index=True)
    cpf_cnpj_enc: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)  # Fernet encrypted
    email: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    telefone: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    limite_credito_centavos: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[StatusCliente] = mapped_column(Enum(StatusCliente), default=StatusCliente.ATIVO)
    criado_em: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    pedidos: Mapped[list["Pedido"]] = relationship("Pedido", back_populates="cliente")


class Produto(Base):
    __tablename__ = "produtos"

    id: Mapped[int] = mapped_column(primary_key=True)
    codigo: Mapped[str] = mapped_column(String(50), unique=True, index=True)
    descricao: Mapped[str] = mapped_column(String(255))
    ncm: Mapped[Optional[str]] = mapped_column(String(8), nullable=True)
    unidade: Mapped[str] = mapped_column(String(6), default="UN")
    custo_medio_centavos: Mapped[int] = mapped_column(Integer, default=0)   # CMP — NUNCA float
    estoque_atual: Mapped[int] = mapped_column(Integer, default=0)          # em unidades inteiras
    estoque_minimo: Mapped[int] = mapped_column(Integer, default=0)
    permite_negativo: Mapped[bool] = mapped_column(Boolean, default=False)
    ativo: Mapped[bool] = mapped_column(Boolean, default=True)

    itens: Mapped[list["PedidoItem"]] = relationship("PedidoItem", back_populates="produto")
    movimentacoes: Mapped[list["MovEstoque"]] = relationship("MovEstoque", back_populates="produto")


class Pedido(Base):
    __tablename__ = "pedidos"

    id: Mapped[int] = mapped_column(primary_key=True)
    cliente_id: Mapped[int] = mapped_column(ForeignKey("clientes.id"))
    data: Mapped[date] = mapped_column(Date)
    status: Mapped[StatusPedido] = mapped_column(Enum(StatusPedido), default=StatusPedido.RASCUNHO)
    total_centavos: Mapped[int] = mapped_column(Integer, default=0)
    observacao: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    criado_em: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    faturado_em: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)

    cliente: Mapped["Cliente"] = relationship("Cliente", back_populates="pedidos")
    itens: Mapped[list["PedidoItem"]] = relationship("PedidoItem", back_populates="pedido", cascade="all, delete-orphan")


class PedidoItem(Base):
    __tablename__ = "pedido_itens"

    id: Mapped[int] = mapped_column(primary_key=True)
    pedido_id: Mapped[int] = mapped_column(ForeignKey("pedidos.id"))
    produto_id: Mapped[int] = mapped_column(ForeignKey("produtos.id"))
    qtd: Mapped[int] = mapped_column(Integer)
    preco_unitario_centavos: Mapped[int] = mapped_column(Integer)

    pedido: Mapped["Pedido"] = relationship("Pedido", back_populates="itens")
    produto: Mapped["Produto"] = relationship("Produto", back_populates="itens")


class MovEstoque(Base):
    __tablename__ = "mov_estoque"

    id: Mapped[int] = mapped_column(primary_key=True)
    produto_id: Mapped[int] = mapped_column(ForeignKey("produtos.id"))
    tipo: Mapped[TipoMovEstoque] = mapped_column(Enum(TipoMovEstoque))
    qtd: Mapped[int] = mapped_column(Integer)
    custo_unitario_centavos: Mapped[int] = mapped_column(Integer, default=0)
    referencia_tipo: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)  # "pedido", "nfe", etc.
    referencia_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    data: Mapped[date] = mapped_column(Date)
    criado_em: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    produto: Mapped["Produto"] = relationship("Produto", back_populates="movimentacoes")

File: erp/frontend/src/api/vendas.ts
import axios from 'axios'
import { formatReal } from './financeiro'
export { formatReal }

const api = axios.create({ baseURL: '/api' })
api.interceptors.request.use(cfg => {
  const token = localStorage.getItem('access_token')
  if (token) cfg.headers.Authorization = `Bearer ${token}`
  return cfg
})

export type TipoPessoa = 'fisica' | 'juridica'
export type StatusCliente = 'ativo' | 'inativo' | 'bloqueado'
export type StatusPedido = 'rascunho' | 'confirmado' | 'faturado' | 'cancelado'

export interface Cliente {
  id: number; tipo: TipoPessoa; nome: string
  email: string | null; telefone: string | null
  limite_credito_centavos: number; status: StatusCliente
}

export interface Produto {
  id: number; codigo: string; descricao: string; unidade: string
  ncm: string | null; custo_medio_centavos: number
  estoque_atual: number; estoque_minimo: number
  permite_negativo: boolean; ativo: boolean
}

export interface PedidoItem {
  id: number; produto_id: number; qtd: number; preco_unitario_centavos: number
}

export interface Pedido {
  id: number; cliente_id: number; data: string; status: StatusPedido
  total_centavos: number; observacao: string | null; itens: PedidoItem[]
}

export interface PosicaoEstoque {
  produto_id: number; codigo: string; descricao: string
  estoque_atual: number; estoque_minimo: number
  custo_medio_centavos: number; alerta_minimo: boolean
}

export interface MovEstoque {
  id: number; produto_id: number; tipo: string; qtd: number
  custo_unitario_centavos: number; data: string
  referencia_tipo: string | null; referencia_id: number | null
}

export const clientesApi = {
  listar: (apenas_ativos = true) => api.get<Cliente[]>('/clientes', { params: { apenas_ativos } }).then(r => r.data),
  criar: (data: Pick<Cliente, 'tipo' | 'nome' | 'email' | 'telefone' | 'limite_credito_centavos'>) =>
    api.post<Cliente>('/clientes', data).then(r => r.data),
}

export const produtosApi = {
  listar: (apenas_ativos = true) => api.get<Produto[]>('/produtos', { params: { apenas_ativos } }).then(r => r.data),
  criar: (data: Pick<Produto, 'codigo' | 'descricao' | 'unidade' | 'ncm' | 'estoque_minimo' | 'permite_negativo'>) =>
    api.post<Produto>('/produtos', data).then(r => r.data),
  entradaEstoque: (data: { produto_id: number; qtd: number; custo_unitario_centavos: number; data: string }) =>
    api.post<Produto>('/estoque/entrada', data).then(r => r.data),
}

export const estoqueApi = {
  posicao: () => api.get<PosicaoEstoque[]>('/estoque/posicao').then(r => r.data),
  movimentacoes: () => api.get<MovEstoque[]>('/estoque/movimentacoes').then(r => r.data),
}

export const pedidosApi = {
  listar: (params?: { cliente_id?: number; status?: StatusPedido }) =>
    api.get<Pedido[]>('/pedidos', { params }).then(r => r.data),
  criar: (data: { cliente_id: number; data: string; itens: Omit<PedidoItem, 'id'>[]; observacao?: string }) =>
    api.post<Pedido>('/pedidos', data).then(r => r.data),
  faturar: (id: number) => api.patch<Pedido>(`/pedidos/${id}/faturar`).then(r => r.data),
}
